python/dask_cudf/dask_cudf/batcher_sortnet.py

```diff
@@ -137,3 +146,164 @@ def sort_delayed_frame(parts, by):
     valid = compute(valid_ct)[0]
     validparts = parts[:valid]
     return validparts
+
+
+def set_partitions_hash(df, columns, npartitions):
+    c = hash_object_dispatch(df[columns], index=False)
+    return np.mod(c, npartitions)
+
+
+def _shuffle_group_2(df, columns, ignore_index):
+    if not len(df):
+        return {}, df
+    ind = hash_object_dispatch(df[columns], index=False)
+    n = ind.max() + 1
+    result2 = group_split_dispatch(
+        df, ind.values.view(), n, ignore_index=ignore_index
+    )
+    return result2, df.iloc[:0]
+
+
+def _shuffle_group_get(g_head, i):
+    g, head = g_head
+    if i in g:
+        return g[i]
+    else:
+        return head
+
+
+def _shuffle_group(df, columns, stage, k, npartitions, ignore_index):
+    c = hash_object_dispatch(df[columns], index=False)
+    typ = np.min_scalar_type(npartitions * 2)
+    c = np.mod(c, npartitions).astype(typ, copy=False)
+    np.floor_divide(c, k ** stage, out=c)
+    np.mod(c, k, out=c)
+    return group_split_dispatch(
+        df, c.astype(np.int32), k, ignore_index=ignore_index
+    )
+
+
+def rearrange_by_hash(
+    df, columns, npartitions, max_branch=None, ignore_index=True
+):
+    n = df.npartitions
+    if max_branch is False:
+        stages = 1
+    else:
+        max_branch = max_branch or 32
+        stages = int(math.ceil(math.log(n) / math.log(max_branch)))
+
+    if stages > 1:
+        k = int(math.ceil(n ** (1 / stages)))
+    else:
+        k = n
+
+    if isinstance(columns, str):
+        columns = [columns]
+    elif isinstance(columns, tuple):
+        columns = list(columns)
+
+    groups = []
+    splits = []
+    joins = []
+
+    inputs = [
+        tuple(digit(i, j, k) for j in range(stages))
+        for i in range(k ** stages)
+    ]
+
+    token = tokenize(df, columns, max_branch)
+
+    start = {
+        ("shuffle-join-" + token, 0, inp): (df._name, i)
+        if i < df.npartitions
+        else df._meta
+        for i, inp in enumerate(inputs)
+    }
+
+    for stage in range(1, stages + 1):
+        group = {  # Convert partition into dict of dataframe pieces
+            ("shuffle-group-" + token, stage, inp): (
+                _shuffle_group,
+                ("shuffle-join-" + token, stage - 1, inp),
+                columns,
+                stage - 1,
+                k,
+                n,
+                ignore_index,
+            )
+            for inp in inputs
+        }
+
+        split = {  # Get out each individual dataframe piece from the dicts
+            ("shuffle-split-" + token, stage, i, inp): (
+                getitem,
+                ("shuffle-group-" + token, stage, inp),
+                i,
+            )
+            for i in range(k)
+            for inp in inputs
+        }
+
+        join = {  # concatenate those pieces together, with their friends
```

Comment:
Would suggest changing the name of this from `join` to `concat` or something similar. This isn't a join :)